### ISSUE TITLE
MDS server configuration is plain wrong

### ISSUE THREAD

**Issue by khenderick** [openvstorage/framework#950]
MDS server configuration is plain wrong

**Comment by khenderick** [comment]
Fixed by #957, packaged in openvstorage-2.7.3-rev.4024.f6b9324

**Comment by JeffreyDevloo** [comment]
### Steps
- Installed latest cluster
- Ran testing code

#### Testing code

The following code will test for: 
- Duplicate paths (issue)
- Files that aren't present
- Amount of mds that are deployed

```
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.
import unittest
import os
from ci.helpers.api import OVSClient
from ovs.dal.lists.vdisklist import VDiskList
from ovs.dal.lists.vpoollist import VPoolList
from ovs.dal.lists.storagerouterlist import StorageRouterList
from ovs.extensions.generic.remote import remote
from ovs.extensions.generic.configuration import Configuration
from ovs.extensions.storageserver.storagedriver import StorageDriverConfiguration


from ovs.lib.mdsservice import MDSServiceController


class VDiskHelper(object):
    @staticmethod
    def create_vdisk(vdisk_name, vpool_name, size, storagerouter_ip, api, timeout=30):
        """
        Create a new vDisk on a certain vPool/storagerouter

        :param vdisk_name: location of a vdisk on a vpool
                           (e.g. /mnt/vpool/test.raw = test.raw, /mnt/vpool/volumes/test.raw = volumes/test.raw )
        :type vdisk_name: str
        :param vpool_name: name of a existing vpool
        :type vpool_name: str
        :param size: size of the new vdisk in BYTES (e.g 10737418240 = 10G)
        :type size: int
        :param storagerouter_ip: ip address of a existing storagerouter
        :type storagerouter_ip: str
        :param api: specify a valid api connection to the setup
        :type api: ci.helpers.api.OVSClient
        :param timeout: time to wait for the task to complete
        :type timeout: int
        :param vpool_name: name of a existing vpool
        :type vpool_name: str
        :return: if success
        :rtype: bool
        """
        vpool_guid = VPoolHelper.get_vpool_by_name(vpool_name).guid
        storagerouter_guid = StoragerouterHelper.get_storagerouter_by_ip(storagerouter_ip).guid

        # remove .raw or .vmdk if is present
        if '.raw' in vdisk_name or '.vmdk' in vdisk_name:
            official_vdisk_name = vdisk_name.split('.')[0]
        else:
            official_vdisk_name = vdisk_name

        data = {"name": official_vdisk_name,
                "size": int(size),
                "vpool_guid": vpool_guid,
                "storagerouter_guid": storagerouter_guid}

        task_guid = api.post(
            api='/vdisks/',
            data=data
        )
        task_result = api.wait_for_task(task_id=task_guid, timeout=timeout)

        if not task_result[0]:
            error_msg = "Creating vdisk `{0}` on vPool `{1}` on storagerouter `{2}` has failed with error {3}" \
                .format(vdisk_name, vpool_name, storagerouter_ip, task_result[1])
            raise RuntimeError(error_msg)
        else:
            print "Creating vdisk `{0}` on vPool `{1}` on storagerouter `{2}` should have succeeded".format(vdisk_name,
                                                                                                            vpool_name,
                                                                                                            storagerouter_ip)
            return task_result[1]

    @staticmethod
    def remove_vdisk(vdisk_guid):
        """
        Remove a vdisk from a vPool

        :param vdisk_guid: guid of a existing vdisk
        :type vdisk_guid: str
        :return: if success
        :rtype: bool
        """


class VPoolHelper(object):
    @staticmethod
    def get_vpool_by_name(vpool_name):
        """
        Get a vpool by name

        :param vpool_name: name of a existing vpool
        :type vpool_name: str
        :return: a vpool object
        :rtype: ovs.dal.hybrids.vpool
        """

        vpool = VPoolList.get_vpool_by_name(vpool_name)
        if vpool:
            return vpool
        else:
            raise RuntimeError("vPool with name `{0}` was not found!".format(vpool_name))


class StoragerouterHelper(object):
    @staticmethod
    def get_storagerouter_by_ip(storagerouter_ip):
        """

        :param storagerouter_ip: ip of a storagerouter
        :type storagerouter_ip: str
        :return: storagerouter object
        :rtype: ovs.dal.hybrids.storagerouter.StorageRouter
        """
        return StorageRouterList.get_by_ip(storagerouter_ip)


class MDSTest(object):
    api = OVSClient('10.100.199.151', 'admin', 'admin')
    # Simple manual mapping for testing
    nodes = {
        "node1": {"guid": "0773a2bb-93fe-4a3d-b4ab-ccdd356ef2d4",
                  "ip": "10.100.199.151"
                  },
        "node2": {"guid": "75219e4b-05b0-4608-a97a-275d7f3d364e",
                  "ip": "10.100.199.152"
                  },
        "node3": {"guid": "6ced823f-c12d-47e1-ab23-e64c0c32b6f6",
                  "ip": "10.100.199.153"
                  }
    }
    vpool_name = "myvpool01"
    # 1G sizes
    size = (1024 * 1024)/2

    @staticmethod
    def run():
        # MDSTest.test_mds()
        MDSTest.check_mds_info()

    @staticmethod
    def test_mds(number_of_vdisks=101):
        # Calculate the real amount of vdisks
        # Add vdisks
        for i in xrange(1, number_of_vdisks + 1):
            vdisk_name = "myvdisk{0}".format(str(i).zfill(3))
            VDiskHelper.create_vdisk(vdisk_name, MDSTest.vpool_name, MDSTest.size, MDSTest.nodes["node1"]["ip"], MDSTest.api)
        # Do a mds checkup to create the mds
        MDSServiceController.mds_checkup()

    @staticmethod
    def _map_mds_path():
        mapping = {}
        for vpool in VPoolList.get_vpools():
            for sd in vpool.storagedrivers:
                for p in sd.partitions:
                    if 'MDS' in p.sub_role:
                        mapping[sd.storage_ip] = {
                            vpool.name: {"mds_mountpoint": p.partition.folder}
                        }
        return mapping

    @staticmethod
    def check_mds_info():
        test_success = True
        mds_keys = ["db_directory", "scratch_directory"]

        for vpool in VPoolList.get_vpools():
            for sd in vpool.storagedrivers:

                with remote(sd.storage_ip, [os]) as rm:
                    config = StorageDriverConfiguration('storagedriver', vpool.guid, sd.storagedriver_id)
                    contents = Configuration.get(config.key)
                    mds_nodes = contents.get('metadata_server', {}).get('mds_nodes', [])
                    # Currently a mds is setup for every 100 vdisks
                    mds_amount = len(vpool.vdisks) / 100 + 1
                    if len(mds_nodes) != mds_amount:
                        print "Found incorrect amount of mds for vpool '{0}' on node '{1}'. Expected '{2}'" \
                              " and got '{3}'".format(vpool.name, sd.storage_ip, mds_amount, len(mds_nodes))
                        test_success = False

                    found_paths = {}
                    for key in mds_keys:
                        found_paths[key] = []
                    # Check paths in mds for overlapping ones
                    for index, mds_node in enumerate(mds_nodes):
                        if mds_node["host"] == sd.storage_ip:
                            for key in mds_keys:
                                if mds_node[key] in found_paths[key]:
                                    print "Certain pointers are overlapping in the volumedrivers configuration" \
                                          " for vpool '{0}' on node '{1}'. Found '{2}' multiple times for entry '{3}'".format(vpool.name, sd.storage_ip, mds_node[key], key)
                                    test_success = False
                                else:
                                    found_paths[key] = mds_node[key]

                    # Check if paths are present on the device
                    for key in mds_keys:
                        fname = found_paths[key]
                        if rm.os.path.isdir(fname) is False:
                            print "The following directory is missing '{0}' for vpool '{1}' on node '{2}'. " \
                                  "The volumedriver might have errored during the creation of this file.".format(fname, vpool.name, sd.storage_ip)
                            test_success = False

        if test_success is True:
            print "MDS test succesful"

if __name__ == "__main__":
    MDSTest().run()
```

### Output

```
{u'status': u'STARTED', u'successful': False, u'failed': False, u'result': u"{'hostname': 'celery@ovs-node-1', 'pid': 7535}", u'ready': False, u'id': u'8e0aa3e5-273d-499c-b237-202fc8eddeab'}
Creating vdisk `myvdisk101` on vPool `myvpool01` on storagerouter `10.100.199.151` should have succeeded
2016-11-08 12:01:44 52800 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 0 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi - Setting initial value for key ovs_ensure_single_ovs.mds.mds_checkup
2016-11-08 12:01:44 53400 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 1 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi - New task ovs.mds.mds_checkup with default params scheduled for execution
2016-11-08 12:01:44 53400 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 2 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi - Amount of jobs pending for key ovs_ensure_single_ovs.mds.mds_checkup: 1
2016-11-08 12:01:44 53500 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 3 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi -   KWARGS: {}
2016-11-08 12:01:44 53700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 4 - INFO - MDS checkup - Started
2016-11-08 12:01:44 53900 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 5 - INFO - MDS checkup - vPool myvpool01
2016-11-08 12:01:45 35700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 6 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-2 - ONLINE
2016-11-08 12:01:46 13100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 7 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-3 - ONLINE
2016-11-08 12:01:46 15400 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 8 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-1 - ONLINE
2016-11-08 12:01:46 15900 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 9 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-1 - Capacity available: False
2016-11-08 12:01:47 36200 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 10 - INFO - Applying storagedriver configuration changes on 10.100.199.151
2016-11-08 12:01:47 36200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 10 - INFO - Applying storagedriver configuration changes on 10.100.199.151
2016-11-08 12:01:48 94700 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 11 - INFO - Changes applied
2016-11-08 12:01:48 94700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 11 - INFO - Changes applied
2016-11-08 12:01:48 94800 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 12 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:48 94800 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 12 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:49 01500 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 13 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-3 - Capacity available: False
2016-11-08 12:01:50 19900 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 14 - INFO - Applying storagedriver configuration changes on 10.100.199.153
2016-11-08 12:01:50 19900 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 14 - INFO - Applying storagedriver configuration changes on 10.100.199.153
2016-11-08 12:01:52 01000 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 15 - INFO - Changes applied
2016-11-08 12:01:52 01000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 15 - INFO - Changes applied
2016-11-08 12:01:52 01000 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 16 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:52 01000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 16 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:52 03100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 17 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-2 - Capacity available: False
2016-11-08 12:01:53 33000 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 18 - INFO - Applying storagedriver configuration changes on 10.100.199.152
2016-11-08 12:01:53 33000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 18 - INFO - Applying storagedriver configuration changes on 10.100.199.152
2016-11-08 12:01:55 25100 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 19 - INFO - Changes applied
2016-11-08 12:01:55 25100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 19 - INFO - Changes applied
2016-11-08 12:01:55 25100 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 20 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:55 25100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 20 - WARNING - Following changes were not applied: mds_nodes
2016-11-08 12:01:57 93000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 21 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-1 - Storing default MDS configuration: [{'host': u'10.100.199.151', 'port': 26301}]
2016-11-08 12:01:57 93300 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 22 - INFO - Applying storagedriver configuration changes on 10.100.199.151
2016-11-08 12:01:57 93300 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 22 - INFO - Applying storagedriver configuration changes on 10.100.199.151
2016-11-08 12:01:59 49100 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 23 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.151:26300]" to "[mds://10.100.199.151:26301]"
2016-11-08 12:01:59 49100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 23 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.151:26300]" to "[mds://10.100.199.151:26301]"
2016-11-08 12:01:59 49200 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 24 - INFO - Changes applied
2016-11-08 12:01:59 49200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 24 - INFO - Changes applied
2016-11-08 12:01:59 49900 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 25 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-3 - Storing default MDS configuration: [{'host': u'10.100.199.153', 'port': 26301}]
2016-11-08 12:01:59 50400 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 26 - INFO - Applying storagedriver configuration changes on 10.100.199.153
2016-11-08 12:01:59 50400 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 26 - INFO - Applying storagedriver configuration changes on 10.100.199.153
2016-11-08 12:02:01 50100 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 27 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.152:26300]" to "[mds://10.100.199.153:26301]"
2016-11-08 12:02:01 50100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 27 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.152:26300]" to "[mds://10.100.199.153:26301]"
2016-11-08 12:02:01 50100 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 28 - INFO - Changes applied
2016-11-08 12:02:01 50100 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 28 - INFO - Changes applied
2016-11-08 12:02:01 50600 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 29 - INFO - MDS checkup - vPool myvpool01 - Storage Router ovs-node-2 - Storing default MDS configuration: [{'host': u'10.100.199.152', 'port': 26301}]
2016-11-08 12:02:01 51000 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 30 - INFO - Applying storagedriver configuration changes on 10.100.199.152
2016-11-08 12:02:01 51000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 30 - INFO - Applying storagedriver configuration changes on 10.100.199.152
2016-11-08 12:02:03 19700 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 31 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.151:26300]" to "[mds://10.100.199.152:26301]"
2016-11-08 12:02:03 19700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 31 - INFO - Changed fs_metadata_backend_mds_nodes from "[mds://10.100.199.151:26300]" to "[mds://10.100.199.152:26301]"
2016-11-08 12:02:03 19700 +0100 - ovs-node-1 - 3470/140436285839104 - extensions/storagedriver - 32 - INFO - Changes applied
2016-11-08 12:02:03 19700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/storagedriver - 32 - INFO - Changes applied
2016-11-08 12:02:03 19700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 33 - INFO - MDS checkup - vPool myvpool01 - Ensuring safety for all virtual disks
2016-11-08 12:02:03 21000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 34 - DEBUG - MDS safety: vDisk 00af5e90-d222-438c-b9af-e1dec8f8162f: Start checkup for virtual disk myvdisk031
2016-11-08 12:02:03 27500 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 35 - DEBUG - MDS safety: vDisk 00af5e90-d222-438c-b9af-e1dec8f8162f: Reconfiguration required. Reasons:
2016-11-08 12:02:03 27500 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 36 - DEBUG - MDS safety: vDisk 00af5e90-d222-438c-b9af-e1dec8f8162f:    * Master overloaded
2016-11-08 12:02:05 03700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 37 - DEBUG - MDS safety: vDisk 00af5e90-d222-438c-b9af-e1dec8f8162f: Completed
2016-11-08 12:02:05 03900 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 38 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572: Start checkup for virtual disk myvdisk101
2016-11-08 12:02:05 09200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 39 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572: Reconfiguration required. Reasons:
2016-11-08 12:02:05 09200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 40 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572:    * Not enough safety
2016-11-08 12:02:05 09200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 41 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572:    * Master overloaded
2016-11-08 12:02:05 09200 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 42 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572:    * Not enough services in use in primary domain
2016-11-08 12:02:06 86700 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 43 - DEBUG - MDS safety: vDisk 01ee9d4d-0163-40fc-8fc6-a9b6298c6572: Completed
....
2016-11-08 12:05:08 09000 +0100 - ovs-node-1 - 3470/140436285839104 - lib/mds - 490 - INFO - MDS checkup - Finished
2016-11-08 12:05:08 31400 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 491 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi - Task ovs.mds.mds_checkup finished successfully
2016-11-08 12:05:08 31500 +0100 - ovs-node-1 - 3470/140436285839104 - lib/ensure single - 492 - INFO - Ensure single CHAINED mode - ID 1478602904_MXcVV2IPSi - Amount of jobs pending for key ovs_ensure_single_ovs.mds.mds_checkup: 0
MDS test succesful
```

### Test result

Test passed.
